samcli/lib/pipeline/bootstrap/stage_resources.yaml

```diff
@@ -0,0 +1,307 @@
+              AWS:
+                - Fn::If:
+                  - MissingPipelineExecutionRole
+                  - !GetAtt PipelineExecutionRole.Arn
+                  - !Ref PipelineExecutionRoleArn
+                - Fn::If:
+                  - MissingCloudFormationExecutionRole
+                  - !GetAtt CloudFormationExecutionRole.Arn
+                  - !Ref CloudFormationExecutionRoleArn
+
+  PipelineExecutionRolePermissionPolicy:
+    Type: AWS::IAM::Policy
+    Condition: MissingPipelineExecutionRole
+    Properties:
+      PolicyName: PipelineExecutionRolePermissions
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Action: 'iam:PassRole'
+            Resource:
+              Fn::If:
+                - MissingCloudFormationExecutionRole
+                - !GetAtt CloudFormationExecutionRole.Arn
+                - !Ref CloudFormationExecutionRoleArn
+          - Effect: Allow
+            Action:
+              - "cloudformation:CreateChangeSet"
+              - "cloudformation:DescribeChangeSet"
+              - "cloudformation:ExecuteChangeSet"
+              - "cloudformation:DescribeStackEvents"
+              - "cloudformation:DescribeStacks"
+              - "cloudformation:GetTemplateSummary"
+              - "cloudformation:DescribeStackResource"
+            Resource: '*'
+          - Effect: Allow
+            Action:
+              - 's3:GetObject*'
+              - 's3:PutObject*'
+              - 's3:GetBucket*'
+              - 's3:List*'
+            Resource:
+              Fn::If:
+                - MissingArtifactsBucket
+                - - !Join [ '',[ !GetAtt ArtifactsBucket.Arn, '/*' ] ]
+                  - !GetAtt ArtifactsBucket.Arn
+                - - !Join [ '',[ !Ref ArtifactsBucketArn, '/*' ] ]
+                  - !Ref ArtifactsBucketArn
+          - Fn::If:
+            - EcrRepoRequired
+            - Effect: "Allow"
+              Action: "ecr:GetAuthorizationToken"
+              Resource: "*"
+            - !Ref AWS::NoValue
+          - Fn::If:
+            - EcrRepoRequired
+            - Effect: "Allow"
+              Action:
+                - "ecr:GetDownloadUrlForLayer"
+                - "ecr:BatchGetImage"
+                - "ecr:BatchCheckLayerAvailability"
+                - "ecr:PutImage"
+                - "ecr:InitiateLayerUpload"
+                - "ecr:UploadLayerPart"
+                - "ecr:CompleteLayerUpload"
+              Resource:
+                Fn::If:
+                  - MissingEcrRepo
+                  - !GetAtt EcrRepo.Arn
+                  - !Ref EcrRepoArn
+            - !Ref AWS::NoValue
+      Roles:
+        - !Ref PipelineExecutionRole
+
+  EcrRepo:
+    Type: AWS::ECR::Repository
+    Condition: MissingEcrRepo
+    Properties:
+      RepositoryPolicyText:
+        Version: "2012-10-17"
+        Statement:
+          - Sid: AllowPushPull
+            Effect: Allow
+            Principal:
```

Comment:
How does Lambda get permission to pull and use the images?

Reply:
To use image-based lambdas, it is the IAM user/role that requires ECR permissions, not the function itself https://docs.aws.amazon.com/lambda/latest/dg/configuration-images.html#configuration-images-permissions